crates/circuit/src/dag_circuit.rs

```diff
@@ -6383,10 +6453,24 @@ impl DAGCircuit {
         &self.op_names
     }
 
-    /// Extends the DAG with valid instances of [PackedInstruction]
+    /// Extends the DAG with valid instances of [PackedInstruction].
     pub fn extend<I>(&mut self, py: Python, iter: I) -> PyResult<Vec<NodeIndex>>
     where
         I: IntoIterator<Item = PackedInstruction>,
+    {
+        self.try_extend(
+            py,
+            iter.into_iter()
+                .map(|inst| -> Result<PackedInstruction, Infallible> { Ok(inst) }),
+        )
+    }
+
+    /// Extends the DAG with valid instances of [PackedInstruction], where the iterator produces the
+    /// results in a fallible manner.
+    pub fn try_extend<I, E>(&mut self, py: Python, iter: I) -> PyResult<Vec<NodeIndex>>
```

Comment:
Thank you for separating the extend method to be able to throw its own type of `Error` result. Do you think this is a pattern that we should be following with fallible code from now on?

Reply:
It depends - the downside this form that I've moved it into is that the DAG gets partially constructed before a potential failure, so it's not easily possible to recover from.  The upside (and why I did it) is that there's no longer a need to collect into an intermediate `Vec` before extending.

I thought that it made sense to do here, because whenever Python's involved, we typically can't _ever_ be certain that things will go off infallibly, and we already have lots of DAG mutation methods that leave the DAG in a partially mutated state if an exception occurs.